### PR TITLE
Add registration of custom types in config

### PR DESCRIPTION
**Pull request by juliusstoerrle**
Adds registration of custom types via configuration, similar to Doctrine ORM.

See Issue: #75

Still need to update config checks and test the code, therefore not yet ready for review.

### CODE REVIEW

**Comment by alcaeus** [comment]
@juliusstoerrle what's the status here? Do you need help or guidance finishing this PR?

**Comment by juliusstoerrle** [comment]
I got a pretty stable version locally now. I will try to finish everything I'm able to fo myself tomorrow and then we can see. I guess there will be some things that could be improved or need to be added.

**Comment by juliusstoerrle** [comment]
So overall this now seems to work. I initially only registered the types in the `Bundle::boot()`-Method but that resulted in errors during container build time. Only registering during build time obviously doesn't work as that code isn't executed in production. Result of that situation is some code duplication, namely the `registerTypes()` functions in the `Bundle` and the `HydratorCacheWarmer` classes. If someone has a better solution I'm happy to change that.

After some research I added a type element to the xml schema, but as I don't use XML config I did not test it. Is there some easy way to test those changes? (the existing unit tests pass)

I did not yet include any changes to the docs, would you like those in this PR or in a separate one?

**Comment by juliusstoerrle** [comment]
To register a custom type when using **yaml** config:

1.  Add a `types:` key to your `doctrine_mongodb` yaml config
2.  Add your type as `key: value` (`name: FQCN`) pair to the `types` key, like below:
```
doctrine_mongodb:
    ...
    types:
        entity_name: 'App\Infrastructure\Doctrine\ODM\Type\EntityNameType'
        form_description: 'App\Infrastructure\Doctrine\ODM\Type\FormDescriptionType'
```
If you use **XML** config, then add for each type an Element like this to your config:
`<doctrine:type name="foo_bar">Bar\FooBarType</doctrine:type>`

File: CacheWarmer/HydratorCacheWarmer.php [hunk 1]
```diff
@@ -66,4 +70,21 @@ public function warmUp($cacheDir)
             $dm->getHydratorFactory()->generateHydratorClasses($classes);
         }
     }
+
+    private function registerTypes() : void
+    {
+        if (! $this->container->hasParameter('doctrine_mongodb.odm.types')) {
+            // no types defined
+            exit;
```
Comment (alcaeus) [inline CacheWarmer/HydratorCacheWarmer.php]: I'm pretty sure this should be a `return`, not an `exit` ;)
```suggestion
            return;
```

File: CacheWarmer/HydratorCacheWarmer.php [hunk 2]
```diff
@@ -44,6 +45,9 @@ public function isOptional()
 
     public function warmUp($cacheDir)
     {
+        // register custom types during cache warm up to avoid errors while generating hydrator classes
+        $this->registerTypes();
```
Comment (alcaeus) [inline CacheWarmer/HydratorCacheWarmer.php]: Why is the logic from the bundle's `boot` method duplicated here? Shouldn't the bundle be booted, thus types being registered?

Comment (juliusstoerrle) [reply CacheWarmer/HydratorCacheWarmer.php]: I initially only registered the types in the Bundle::boot()-Method but that resulted in errors during container build time. Only registering during build time obviously doesn't work as that code isn't executed in production. Result of that situation is some code duplication, namely the registerTypes() functions in the Bundle and the HydratorCacheWarmer classes. If someone has a better solution I'm happy to change that.

Comment (juliusstoerrle) [reply CacheWarmer/HydratorCacheWarmer.php]: > Shouldn't the bundle be booted, thus types being registered?

Thats what I initally expected aswell.

**Comment by alcaeus** [comment]
Closing in favour of #631.

**Pull request closed without merging**